### PR TITLE
Add reusable workflows and Guide for end to end testing with Uptest

### LINKED ISSUE #31: Create uptest Github Action
### What problem are you facing?

Currently, we are configuring repetitive GHA workflows across the Configurations under test,e.g.

* https://github.com/upbound/platform-ref-gcp/blob/main/.github/workflows/e2e.yaml
* https://github.com/upbound/platform-ref-azure/pull/18
* https://github.com/upbound/platform-ref-aws/pull/82

See initial [contextual discussion](https://github.com/upbound/platform-ref-azure/pull/18#discussion_r998146115)

### How could Upbound help solve your problem?

Encapsulate uptest e2e testing logic into standalone [composite Github Action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) or [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows)

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes

This PR adds reusable workflows to be reused by provider or configuration repositories for end to end testing.
It also adds a step by step guide on how to integrate it.

Fixes #31 

I have:

- [x] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR, as appropriate.

### How has this code been tested

Following the guide at `docs/integrating-uptest-for-e2e-testing.md` that is being introduced with this PR as well.

### CODE REVIEW

**Comment by turkenh** [comment]
As a reference for the reviewer, here is how it looks like after the guide followed step by step: https://github.com/turkenh/demo-uptest-integration

**Comment by ytsarev** [comment]
@turkenh could you please give me access to https://github.com/turkenh/demo-uptest-integration ? I'm getting 404

**Comment by turkenh** [comment]
> @turkenh could you please give me access to https://github.com/turkenh/demo-uptest-integration ? I'm getting 404

just made it public, can you check again?

File: docs/integrating-uptest-for-e2e-testing.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
+# Integrating Uptest for End to End Testing
+
+In this tutorial, we will integrate [uptest](https://github.com/upbound/uptest) to a Github repository to automate end to end
+testing managed resources. While we will use a _provider_ repository as an example, the process will be identical for a
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:4]:
```suggestion
testing managed resources. While we will use a `Provider` repository as an example, the process will be identical for a
```

File: docs/integrating-uptest-for-e2e-testing.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+configuration repository.
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:5]:
```suggestion
`Configurarion` repository.
```
Highlight Crossplane packages kinds to not confuse the reader with the arbitrary configuration repo

File: docs/integrating-uptest-for-e2e-testing.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+2. Create a PR with that change.
+3. Add the following comment on the PR:
+
+	```
+	/test-examples=examples/repository/repository.yaml
+	```
+
+4. Check the Actions and follow how end to end testing goes.
+
+[demo repository]: (https://github.com/upbound/demo-uptest-integration)
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:143]:
```suggestion
[demo repository]: https://github.com/upbound/demo-uptest-integration
```
Link was not working while in parenthesis

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    types: [created]
+
+jobs:
+  job1:
+    uses: ./.github/workflows/pr-comment-trigger.yml
+    secrets:
+      UPTEST_CLOUD_CREDENTIALS: ${{ secrets.UPTEST_AWS_CREDS }}
+      UPTEST_DATASOURCE: ${{ secrets.UPTEST_EXAMPLE_VALUE_REPLACEMENTS }}
+    #with:
+      #trigger-keyword: "/test-e2e"
```
Comment (ytsarev) [inline .github/workflows/test.yml:14]: is it still required?

File: docs/integrating-uptest-for-e2e-testing.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	on:
+	  issue_comment:
+	    types: [created]
+
+	jobs:
+	  e2e:
+	    uses: upbound/uptest/.github/workflows/pr-comment-trigger.yml@main
+	    secrets:
+	      UPTEST_CLOUD_CREDENTIALS: \${{ secrets.UPTEST_AWS_CREDS }}
+	      UPTEST_DATASOURCE: \${{ secrets.UPTEST_EXAMPLE_VALUE_REPLACEMENTS }}
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:110]: `UPTEST_EXAMPLE_VALUE_REPLACEMENTS` is not documented anywhere later in the doc

**Review by ytsarev — COMMENTED** [review]
Left some comments over the doc, still need access to the test/demo repo

File: docs/integrating-uptest-for-e2e-testing.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+testing managed resources. While we will use a _provider_ repository as an example, the process will be identical for a
+configuration repository.
+
+Starting with a provider repository with no end to end testing capability, we will end up having:
+- A make target to locally test examples end to end
+- A GitHub action triggered for PRs whenever a comment as `/test-examples=<examples-path>` is left
+
+## Setting up the Make targets
+
+1. Go to the [demo repository] which contains a GitHub provider generated using upjet and hit the `Use this template` button
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:13]: trying to use https://github.com/turkenh/demo-uptest-integration in this context, looks like this repo is not a template

File: .github/workflows/test.yml (hunk 4 above)
Comment (turkenh) [reply .github/workflows/test.yml:14]: No, good catch.

File: docs/integrating-uptest-for-e2e-testing.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	cat <<EOF > .github/workflows/e2e.yaml
+	name: End to End Testing
+
+	on:
+	  issue_comment:
+	    types: [created]
+
+	jobs:
+	  e2e:
+	    uses: upbound/uptest/.github/workflows/pr-comment-trigger.yml@main
```
Comment (turkenh) [inline docs/integrating-uptest-for-e2e-testing.md:107]: @ytsarev if you want to give it a try before this PR merges, you would need to replace this line as below:

```suggestion
	    uses: turkenh/uptest/.github/workflows/pr-comment-trigger.yml@main
```

**Comment by turkenh** [comment]
@ytsarev this is now ready for another pass. 
Sorry for the confusion due to things being private.

File: docs/integrating-uptest-for-e2e-testing.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	    secrets:
+	      UPTEST_CLOUD_CREDENTIALS: \${{ secrets.UPTEST_CLOUD_CREDENTIALS }}
+	      UPTEST_DATASOURCE: \${{ secrets.UPTEST_DATASOURCE }}
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:110]: `UPTEST_CLOUD_CREDENTIALS` are self-explanatory, but `UPTEST_DATASOURCE` is not obvious to the reader without advanced uptest experience, could we maybe reference the datasource concept in the comment here?

File: docs/integrating-uptest-for-e2e-testing.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+2. Commit and push to the `main` branch of the repository.
+
+	```
+	git add .github/workflows/e2e.yaml
+	git commit -s -m "Add e2e workflow"
+	git push origin main
+	```
+
+3. Lastly, we need to add a Repository Secret with our GitHub token.
+	1. Go to your repository settings in GitHub UI.
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:123]:
```suggestion
	1. Go to your repository Settings in GitHub UI.
```

File: docs/integrating-uptest-for-e2e-testing.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	2. On the left side, select `Secrets` -> `Actions` under `Security` section.
+	3. Hit `New Repository Secret`
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:125]:
```suggestion
	3. Hit `New repository secret`
```
nit: keep case consistent with GitHub UI

File: docs/integrating-uptest-for-e2e-testing.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+	4. Enter `UPTEST_CLOUD_CREDENTIALS` as `Name` and your GitHub Token as `Secret` and hit `Add secret`
+
+## Testing via CI
+
+We are now ready to test our changes end to end via GitHub Actions. We will try that out by opening a test PR.
+
+1. Go to the `examples/repository/repository.yaml` file and make some wording changes in `description` field, e.g.
```
Comment (ytsarev) [inline docs/integrating-uptest-for-e2e-testing.md:132]: Strictly following the guide, just before the sample file modification, the git status of the repo is not clean:

```
git status --short
 M Makefile
?? cluster/test/
?? kubeconfig
```
So we need to
* Instruct the reader to commit `Makefile` and `cluster/test/setup.sh` somewhere before this step.
* Add `kubconfig` to `.gitignore` of https://github.com/upbound/demo-uptest-integration/blob/main/.gitignore

**Review by ytsarev — CHANGES_REQUESTED** [review]
Went through the whole flow end-to-end with a sample repo, and overall it looks great. 

 Left a couple of comments and suggestions.

File: docs/integrating-uptest-for-e2e-testing.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+## Testing Locally
+
+1. Generate a [Personal Access Token](https://github.com/settings/tokens/new) for your Github account with
+    `repo/public_repo` and `delete_repo` scopes.
+2. Run the following:
+
+	```bash
+	export UPTEST_CLOUD_CREDENTIALS=<your-token-here>
+	UPTEST_EXAMPLE_LIST=examples/repository/repository.yaml make e2e
```
[the remark this answers is not in the document]
Comment (turkenh) [reply docs/integrating-uptest-for-e2e-testing.md:85]: @ytsarev there is no strict go requirement we have here, but this check is coming from build submodule, see relevant discussion: https://github.com/upbound/build/pull/195#issuecomment-1201471605

File: docs/integrating-uptest-for-e2e-testing.md [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+	-include build/makelib/local.xpkg.mk
+	-include build/makelib/controlplane.mk
+
+	uptest: $(UPTEST) $(KUBECTL) $(KUTTL)
+		@$(INFO) running automated tests
+		@KUBECTL=$(KUBECTL) KUTTL=$(KUTTL) $(UPTEST) e2e "${UPTEST_EXAMPLE_LIST}" --setup-script=cluster/test/setup.sh || $(FAIL)
+		@$(OK) running automated tests
+
+	e2e: build controlplane.up local.xpkg.deploy.provider.$(PROJECT_NAME) uptest
+	```
```
[the remark this answers is not in the document]
Comment (turkenh) [reply docs/integrating-uptest-for-e2e-testing.md:75]: Yes, after this PR merges. Just created https://github.com/upbound/upjet-provider-template/issues/14

File: docs/integrating-uptest-for-e2e-testing.md (hunk 11 above)
Comment (turkenh) [reply docs/integrating-uptest-for-e2e-testing.md:143]: Both done!

**Comment by turkenh** [comment]
@ytsarev all your comments resolved, thanks!

**Review by ytsarev — APPROVED** [review]
Thanks a lot for all the fixes, looks great!

**Pull request merged**